accepted/2021/package-validation/apicompat.md

```diff
@@ -0,0 +1,143 @@
+
+Also, this rule should allow differences in nullable annotations that are not breaking, for example, relaxing a nullable reference type annotation on an in parameter (i.e `public string Foo(string a) -> public string Foo(string? a)`). However, this should be configurable via rule settings to disable this as in some cases where you are comparing a reference assembly vs it's implementation you want them to match in all cases.
+
+### 14. Constructor makes base class instantiable.
+Rule that checks if a public API is a constructor that makes a class instantiable, even though the base class is not instantiable. That API pattern is not allowed, because it causes protected members of the base class, which are not considered public APIs, to be exposed to subclasses of this class.
+
+## Configurability
+
+Given that we are proposing a Roslyn analyzer a dotnet tool and potentially an MSBuild Task, we should support a friendly format for all three possible scenarios to baseline warnings/errors, configure rules and severity of rules.
+
+To baseline errors we could support:
+
+1. All analyzer's disable mechanism when using the tool as an analyzer.
+2. A txt file that lists the errors to baseline (maybe we could support this for the analyzer scenario as well).
+
+Configuring rules involves settings like, ignoring certain attributes as describe in the attribute compat rule, rules number 8 in the document, etc. I think these could be part of the `.editorconfig` with configuration for rules' severity.
+
+### MSBuild support
+
+We should provide MSBuild first class configurability via properties so that costumers can just add a `PackageReference` and set properties on how they want to consume the tool (via MSBuild Task, Analyzer, etc), what the inputs are going to be, where the baseline errors file can be found, etc. This would make it very simple for customers to include in their libraries and configure without having to write complicated MSBuild targets to run the tool on an MSBuild process if that is their best way to hook it up into their build system.
+
+With this we could extend support to make it easier for packages to run validation by automatically download previous version(s) of the package and then run compatibility checks vs the current version by just setting a few MSBuild properties.
+
+## Diagnostics consistency
+
+When running as an analyzer's we are going to provide compiler diagnostics, so when running outside of an analyzer context, we should provide diagnostics that are consistent on all scenarios so that baselining these diagnostics and understanding them is clear within an editor, console or msbuild process.
+
+So something like:
+
+```
+error RS0017: <helpful error description that doesn't depend of a squiggle to tell what's going on>
+```
+
+## Open questions
+
+1. Where should these tools live? Maybe [dotnet/roslyn-analyzers](https://github.com/dotnet/roslyn-analyzers)?
```

Comment:
What is the distribution mechanism?  Are the tools part of the .NET SDK? Distributed separately as a nuget package?

Reply:
I'll add that as an open question, but I think it should be distributed separately as a nuget package.